src/data_race.rs

```diff
@@ -0,0 +1,1407 @@
+//! Implementation of a data-race detector
+//!  uses Lamport Timestamps / Vector-clocks
+//!  base on the Dyamic Race Detection for C++:
+//!     - https://www.doc.ic.ac.uk/~afd/homepages/papers/pdfs/2017/POPL.pdf
+//!  to extend data-race detection to work correctly with fences
+//!  and RMW operations
+//! This does not explore weak memory orders and so can still miss data-races
+//!  but should not report false-positives
+
+use std::{fmt::{self, Debug}, cmp::Ordering, rc::Rc, cell::{Cell, RefCell, Ref, RefMut}, ops::Index};
+
+use rustc_index::vec::{Idx, IndexVec};
+use rustc_target::abi::Size;
+use rustc_middle::ty::layout::TyAndLayout;
+use rustc_data_structures::fx::FxHashMap;
+
+use smallvec::SmallVec;
+
+use crate::*;
```

Comment:
```suggestion
use crate::MiriEvalContext;
```
Since there are only a couple of includes using `crate` below?